File: Solutions/Companies/Google/ProductofArrayExceptSelf.java
package Companies.Google;

public class ProductofArrayExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        int[] re = new int[nums.length];
        re[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            re[i] = re[i-1] * nums[i-1];
        }
        int temp = 1;
        for (int i = nums.length-2; i >= 0; i--) {
            temp *= nums[i+1];
            re[i] *= temp;
        }
        return re;
    }
}

File: Companies/VMware/ProductofArrayExceptSelf.cpp
#include <iostream>
#include <vector>

using namespace std;

class ProductofArrayExceptSelf
{
public:
  vector<int> productExceptSelf(vector<int> &nums)
  {
    vector<int> dp(nums.size(), 1);
    for (int i = 1; i < nums.size(); i++)
    {
      dp[i] = dp[i - 1] * nums[i - 1];
    }
    int temp = nums.back();
    for (int i = nums.size() - 2; i >= 0; i--)
    {
      dp[i] *= temp;
      temp *= nums[i];
    }
    return dp;
  }
};
